src/openms/source/CHEMISTRY/ISOTOPEDISTRIBUTION/IsoSpec.cpp

```diff
@@ -61,41 +62,30 @@ namespace OpenMS
   {
   }
 
-  const std::vector<double>& IsoSpec::getMasses() {return masses_;}
-  const std::vector<double>& IsoSpec::getProbabilities() {return probabilities_;}
-
-  void IsoSpec::run_(Iso* iso)
+  std::vector<Peak1D> IsoSpec::run_(Iso& iso)
   {
     int tabSize = 1000;
     int hashSize = 1000;
 
-    IsoThresholdGenerator* generator = new IsoThresholdGenerator(std::move(*iso), threshold_, absolute_, tabSize, hashSize); 
-    Tabulator<IsoThresholdGenerator>* tabulator = new Tabulator<IsoThresholdGenerator>(generator, true, true, true, true); 
-
-    int size = tabulator->confs_no();
+    IsoThresholdGenerator generator(std::move(iso), threshold_, absolute_, tabSize, hashSize); 
 
-    masses_.clear();
-    masses_.reserve(size);
-    copy(&tabulator->masses()[0], &tabulator->masses()[size], back_inserter(masses_));
+    std::vector<Peak1D> distribution;
```

Comment:
I was thinking about that too, but I dont think we can. Since the loop is over a generator, the configurations are not generated yet and are produced on-the-fly, therefore I dont think we know the size beforehand. Maybe @michalsta can comment here whether I understand this correctly.

Reply:
Yeah, that boils down to estimating the number of configurations above a given threshold in a cartesian product of multinomial distributions. We (me and Mateusz, the other IsoSpec developer) have given quite a lot of thought about that a while ago, we even got desperate enough to search the literature ;) but came up with nothing sensible. Not even an estimate that could be depended upon not to be off by several orders of magnitude in its pessimistic cases. 

So, for now, we have given that up. I think that solving this particular problem might be worth a publication by itself (not Nature obviously ;)

But for now, the only way we can compute the size is to go through the configuration space and count the configurations.

Reply:
...aaand I just checked and actually that seems to be about 20% faster. That is, going through the configuration space just counting configurations w/o storing them, then reserving the proper size of the vector, then throwing away the previous iso, constructing a new one and generating into a pre-allocated vector is on my machine faster than just going through the config space once, storing them in vector, and depending on the vector to grow as needed. That would also have the advantage of leaving the backend storage of the vector at the exact size needed, without the up to 2x overhead.

But this is going to be dependant as hell on the machine memory bandwidth, cache sizes, and most importantly, the efficiency of the OS memory manager and malloc implementation, so for now let's leave it as it is. I'll try to do some more involved tests than just "works for me" (like on other machines and OSes), and make use of this knowledge in the next IsoSpec release (there is obviously lot of optimization on our side possible here, like reusing marginals, parsed formula, and so on for counting and computing). So for now, let's leave it as it is, and when we finalize IsoSpec 1.9.0 I'll make a code dump and change it.